File: src/m3/contrib/kladr/addrfield.py
#coding:utf-8

from django.db.models.query_utils import Q

from m3.ui.ext.containers.base import BaseExtContainer
from m3.ui.ext.fields.simple import ExtHiddenField
from m3.ui.actions import Action, PreJsonResult, OperationResult
from m3.contrib.kladr.models import KladrGeo, KladrStreet
from actions import KLADRPack, kladr_controller

class ExtAddrComponent(BaseExtContainer):
    '''
    Блок указания адреса 
    '''
    PLACE = 1 # Уровнь населенного пункта
    STREET = 2 # Уровнь улицы 
    HOUSE = 3 # Уровень дома
    FLAT = 4 # Уровень квартиры    
    
    VIEW_0 = 0 # хитрый режим (пока не будем делать), когда отображается только адрес, а его редактирование отдельным окном
    VIEW_1 = 1 # в одну строку + адрес
    VIEW_2 = 2 # в две строки + адрес, только для level > PLACE
    VIEW_3 = 3 # в три строки + адрес, только для level > STREET        
    
    def __init__(self, *args, **kwargs):
        super(ExtAddrComponent, self).__init__(*args, **kwargs)
        self.place_field_name = 'place'
        self.street_field_name = 'street'
        self.house_field_name = 'house'
        self.flat_field_name = 'flat'
        self.addr_field_name = 'addr'
        self.place_label = u'Населенный пункт'
        self.street_label = u'Улица'
        self.house_label = u'Дом/Корпус'
        self.flat_label = u'Квартира'
        self.addr_label = u'Адрес'
        
        self.pack = kladr_controller.find_pack(KLADRPack)
        #self.action_getaddr = self.pack.get_addr_action
        self.level = ExtAddrComponent.FLAT
        self.addr_visible = True
        self.view_mode = ExtAddrComponent.VIEW_2
        self.init_component(*args, **kwargs)
        self.layout = 'form'
        self.template = 'ext-fields/ext-addr-field.js'
        self.addr = ExtHiddenField(name = self.addr_field_name, type = ExtHiddenField.STRING);
        self.place = ExtHiddenField(name = self.place_field_name, type = ExtHiddenField.STRING);
        self.street = ExtHiddenField(name = self.street_field_name, type = ExtHiddenField.STRING);
        self.house = ExtHiddenField(name = self.house_field_name, type = ExtHiddenField.STRING);
        self.flat = ExtHiddenField(name = self.flat_field_name, type = ExtHiddenField.STRING);
        self._items.append(self.addr)
        self._items.append(self.place)
        self._items.append(self.street)
        self._items.append(self.house)
        self._items.append(self.flat)
        if self.view_mode == ExtAddrComponent.VIEW_1:
            self.height = 25
        elif self.view_mode == ExtAddrComponent.VIEW_2:
            if self.level > ExtAddrComponent.STREET:
                self.height = 25*2
            else:
                self.height = 25
        elif self.view_mode == ExtAddrComponent.VIEW_3:
            if self.level > ExtAddrComponent.HOUSE:
                self.height = 25*3
            else:
                if self.level > ExtAddrComponent.STREET:
                    self.height = 25*2
                else:
                    self.height = 25
        if self.addr_visible:
            self.height += 36+7

    def render_params(self):
        super(ExtAddrComponent, self).render_params()
        self._put_params_value('place_field_name', self.place_field_name)
        self._put_params_value('street_field_name', self.street_field_name)
        self._put_params_value('house_field_name', self.house_field_name)
        self._put_params_value('flat_field_name', self.flat_field_name)
        self._put_params_value('addr_field_name', self.addr_field_name)
        self._put_params_value('place_label', self.place_label)
        self._put_params_value('street_label', self.street_label)
        self._put_params_value('house_label', self.house_label)
        self._put_params_value('flat_label', self.flat_label)
        self._put_params_value('addr_label', self.addr_label)
        self._put_params_value('addr_visible', (True if self.addr_visible else False ))
        self._put_params_value('level', self.level)
        self._put_params_value('view_mode', self.view_mode)
        self._put_params_value('place_value', (self.place.value if self.place and self.place.value else ''))
        self._put_params_value('place_text', (self.pack.get_place_name(self.place.value) if self.place and self.place.value else ''))
        self._put_params_value('street_value', (self.street.value if self.street and self.street.value else ''))
        self._put_params_value('street_text', (self.pack.get_street_name(self.street.value) if self.street and self.street.value else ''))
        self._put_params_value('house_value', (self.house.value if self.house and self.house.value else ''))
        self._put_params_value('flat_value', (self.flat.value if self.flat and self.flat.value else ''))
        self._put_params_value('addr_value', (self.addr.value if self.addr and self.addr.value else ''))
        self._put_params_value('get_addr_url', (self.pack.get_addr_action.absolute_url() if self.pack.get_addr_action else ''))
        self._put_params_value('kladr_url', (self.pack.get_places_action.absolute_url() if self.pack.get_places_action else ''))
        self._put_params_value('street_url', (self.pack.get_streets_action.absolute_url() if self.pack.get_streets_action else ''))
    
    def render_base_config(self):
        res = super(ExtAddrComponent, self).render_base_config()        
        return res
    
    def render(self):
        self.render_base_config()
        self.render_params()
        
        config = self._get_config_str()
        params = self._get_params_str()
        return 'new Ext.m3.AddrField({%s},{%s})' % (config, params)

    @property
    def items(self):       
        return self._items
    
    @property
    def handler_change_place(self):
        return self._listeners.get('change_place')
    
    @handler_change_place.setter
    def handler_change_place(self, function):
        self._listeners['change_place'] = function
    
    @property
    def handler_change_street(self):
        return self._listeners.get('change_street')
    
    @handler_change_street.setter
    def handler_change_street(self, function):
        self._listeners['change_street'] = function
    
    @property
    def handler_change_house(self):
        return self._listeners.get('change_house')
    
    @handler_change_house.setter
    def handler_change_house(self, function):
        self._listeners['change_house'] = function
    
    @property
    def handler_change_flat(self):
        return self._listeners.get('change_flat')
    
    @handler_change_flat.setter
    def handler_change_flat(self, function):
        self._listeners['change_flat'] = function

    @property
    def handler_before_query_place(self):
        return self._listeners.get('before_query_place')
    
    @handler_before_query_place.setter
    def handler_before_query_place(self, function):
        self._listeners['before_query_place'] = function

File: src/m3/static/m3/js/AddrField.js
/**
 * Панель редактирования адреса
 */
Ext.m3.AddrField = Ext.extend(Ext.Container, {
	constructor: function(baseConfig, params){
		 
		var items = params.items || [];
		
		var place_store = new Ext.data.JsonStore({
			url: params.kladr_url,
			idProperty: 'code',
			root: 'rows',
			totalProperty: 'total',
			fields: [{name: 'code'},
				{name: 'display_name'},
				{name: 'socr'},
				{name: 'zipcode'},
				{name: 'gni'},
				{name: 'uno'},
				{name: 'okato'}
			]
		});
		var record = new Ext.data.Record();
    	record.code = params.place_value;
    	record.display_name = params.place_text;
		place_store.loadData({total:1, rows:[record]});
		
		this.place = new Ext.form.ComboBox({
			name: params.place_field_name,
			fieldLabel: params.place_label,
			hideTrigger: true,
			minChars: 2,
			emptyText: 'Населенный пункт...',
			queryParam: 'filter',
			store: place_store,
			resizable: true,
			displayField: 'display_name',
			valueField: 'code',
			mode: 'remote',
			hiddenName: params.place_field_name
		});		
		this.place.setValue(params.place_value);
		
		if (params.level > 1) {
			var street_store = new Ext.data.JsonStore({
				url: params.street_url,
				idProperty: 'code',
				root: 'rows',
				totalProperty: 'total',
				fields: [{name: 'code'},
					{name: 'display_name'},
					{name: 'socr'},
					{name: 'zipcode'},
					{name: 'gni'},
					{name: 'uno'},
					{name: 'okato'}
				]
			});
			var record = new Ext.data.Record();
			record.code = params.street_value;
			record.display_name = params.street_text;
			street_store.loadData({
				total: 1,
				rows: [record]
			});
			
			this.street = new Ext.form.ComboBox({
				name: params.street_field_name,
				fieldLabel: params.street_label,
				hideTrigger: true,
				minChars: 2,
				emptyText: 'Улица...',
				queryParam: 'filter',
				store: street_store,
				resizable: true,
				displayField: 'display_name',
				valueField: 'code',
				mode: 'remote',
				hiddenName: params.street_field_name
			});
			this.street.setValue(params.street_value);
			
			if (params.level > 2) {
				this.house = new Ext.form.TextField({
					name: params.house_field_name,
					fieldLabel: params.house_label,
					value: params.house_value,
					emptyText: 'Дом...',
					width: 40
				});
				
				if (params.level > 3) {
					this.flat = new Ext.form.TextField({
						name: params.flat_field_name,
						fieldLabel: params.flat_label,
						value: params.flat_value,
						emptyText: 'Кв-ра...',
						width: 40
					});
				}
			}
		};
		if (params.addr_visible) {
			this.addr = new Ext.form.TextArea({
				name: params.addr_field_name,
				anchor: '100%',
				fieldLabel: params.addr_label,
				value: params.addr_value,
				readOnly: true,
				height: 36
			});
		};
		if (params.view_mode == 1){
			// В одну строку
			this.place.flex = 1;
			var row_items = [this.place];
			if (params.level > 1) {
				this.street.flex = 1;
				this.street.fieldLabel = params.place_label;
				row_items.push({
						xtype: 'label'
						,style: {padding:'3px'}
    					,text: params.street_label+':'
					}
					, this.street
				);
				if (params.level > 2) {
					this.house.fieldLabel = params.place_label;
					row_items.push({
							xtype: 'label'
							,style: {padding:'3px'}
	    					,text: params.house_label+':'
						}
						, this.house
					);
					if (params.level > 3) {
						this.flat.fieldLabel = params.place_label;
						row_items.push({
								xtype: 'label'
								,style: {padding:'3px'}
		    					,text: params.flat_label+':'
							}
							, this.flat
						);
					}
				}
			}
			var row = {
				xtype: 'compositefield'
				, anchor: '100%'
				, fieldLabel: params.place_label
				, items: row_items
			};
			items.push(row);
			if (params.addr_visible) {
				items.push(this.addr);
			}
		};
		if (params.view_mode == 2){
			// В две строки
			this.place.anchor = '100%';
			items.push(this.place);
			if (params.level > 1) {
				this.street.flex = 1;
				var row_items = [this.street];
				if (params.level > 2) {
					this.house.fieldLabel = params.street_label;
					row_items.push({
							xtype: 'label'
							,style: {padding:'3px'}
	    					,text: params.house_label+':'
						}
						, this.house
					);
					if (params.level > 3) {
						this.flat.fieldLabel = params.street_label;
						row_items.push({
								xtype: 'label'
								,style: {padding:'3px'}
		    					,text: params.flat_label+':'
							}
							, this.flat
						);
					}
				};
				var row = {
					xtype: 'compositefield'
					, anchor: '100%'
					, fieldLabel: params.street_label
					, items: row_items
				};
				items.push(row);
			}
			if (params.addr_visible) {
				items.push(this.addr);
			}
		};
		if (params.view_mode == 3){
			// В три строки
			this.place.anchor = '100%';
			items.push(this.place);
			if (params.level > 1) {
				this.street.anchor = '100%';
				items.push(this.street);
				if (params.level > 2) {
					var row_items = [{
						xtype: 'container'
						, layout: 'form'
						, items: this.house 
					}];
					if (params.level > 3) {
						row_items.push({
							xtype: 'container'
							, layout: 'form'
							, style: {padding: '0px 0px 0px 5px'}
							, items: this.flat
						});
					}
					var row = new Ext.Container({
						anchor: '100%'
						, layout: 'column'
						, items: row_items
					});
					items.push(row);
				}
			};
			if (params.addr_visible) {
				items.push(this.addr);
			}
		};
						
		var config = Ext.applyIf({
			items: items
			, get_addr_url: params.get_addr_url
			, level: params.level
			, addr_visible: params.addr_visible
		}, baseConfig);
		
		Ext.Container.superclass.constructor.call(this, config);
	}
	, beforeStreetQuery: function(qe) {
		this.street.getStore().baseParams.place_code = this.place.value;		
	}
	, clearStreet: function() {		
    	this.street.setValue('');		
	}
	, initComponent: function(){
		Ext.m3.AddrField.superclass.initComponent.call(this);		
		this.mon(this.place, 'change', this.onChangePlace, this);
		if (this.level > 1) {
			this.mon(this.street, 'change', this.onChangeStreet, this);
			if (this.level > 2) {
				this.mon(this.house, 'change', this.onChangeHouse, this);
				if (this.level > 3) {
					this.mon(this.flat, 'change', this.onChangeFlat, this);
				}
			}
		}
		if (this.level > 1) {
			this.mon(this.place, 'change', this.clearStreet, this);
			this.mon(this.street, 'beforequery', this.beforeStreetQuery, this);
		};
		this.addEvents(
			/**
             * @event change_place
             * При изменении населенного пункта 
             * @param {AddrField} this
             * @param {Place_code} Код нас. пункта по КЛАДР
             * @param {Store} Строка с информацией о данных КЛАДРа по выбранному пункту
             */
			'change_place',
			/**
             * @event change_street
             * При изменении улицы 
             * @param {AddrField} this
             * @param {Street_code} Код улицы по КЛАДР
             * @param {Store} Строка с информацией о данных КЛАДРа по выбранной улице
             */
			'change_street',
			/**
             * @event change_house
             * При изменении дома 
             * @param {AddrField} this
             * @param {House} Номер дома             
             */
			'change_house',
			/**
             * @event change_flat
             * При изменении квартиры 
             * @param {AddrField} this
             * @param {Flat} Номер квартиры             
             */
			'change_flat');
	}	
	, getNewAddr: function (){
		var place_id;
		if (this.place != undefined) {
			place_id = this.place.getValue();
		}
		var street_id;
		if (this.street != undefined) {
			street_id = this.street.getValue();
		}
		var house_num;
		if (this.house != undefined) {
			house_num = this.house.getValue();
		}
		var flat_num;
		if (this.flat != undefined) {
			flat_num = this.flat.getValue();
		}		
		Ext.Ajax.request({
			url: this.get_addr_url,
			params: Ext.applyIf({ place: place_id, street: street_id, house: house_num, flat: flat_num, addr_cmp: this.addr.id }, this.params),
			success: function(response, opts){ smart_eval(response.responseText); },
			failure: function(){Ext.Msg.show({ title:'', msg: 'Не удалось получить адрес.<br>Причина: сервер временно недоступен.', buttons:Ext.Msg.OK, icon: Ext.Msg.WARNING });}
		});
    }
	, setNewAddr: function(newAddr){
		if (this.addr != undefined) {
			this.addr.value = newAddr;
		}
	}
	, onChangePlace: function(){
		var val = this.place.getValue();
		var data =  this.place.getStore().data.get(val);
		if (data != undefined) {
			data = data.data;
		}
		this.fireEvent('change_place', this, val, data);
		if (this.addr_visible) {
			this.getNewAddr();
		}
	}
	, onChangeStreet: function(){
		var val = this.street.getValue();
		var data =  this.street.getStore().data.get(val);
		if (data != undefined) {
			data = data.data;
		}
		this.fireEvent('change_street', this, val, data);
		if (this.addr_visible) {
			this.getNewAddr();
		}
	}
	, onChangeHouse: function(){
		this.fireEvent('change_house', this, this.house.getValue());
		if (this.addr_visible) {
			this.getNewAddr();
		}
	}
	, onChangeFlat: function(){
		this.fireEvent('change_flat', this, this.flat.getValue());
		if (this.addr_visible) {
			this.getNewAddr();
		}
	}
})